services/proxy/pkg/middleware/mfa.go

```diff
@@ -0,0 +1,73 @@
+package middleware
+
+import (
+	"net/http"
+
+	"github.com/owncloud/ocis/v2/ocis-pkg/log"
+	"github.com/owncloud/ocis/v2/ocis-pkg/mfa"
+	"github.com/owncloud/ocis/v2/ocis-pkg/oidc"
+	"github.com/owncloud/ocis/v2/services/proxy/pkg/config"
+)
+
+// MultiFactor returns a middleware that checks requests for mfa
+func MultiFactor(cfg config.MFAConfig, opts ...Option) func(next http.Handler) http.Handler {
+	options := newOptions(opts...)
+	logger := options.Logger
+
+	return func(next http.Handler) http.Handler {
+		return &MultiFactorAuthentication{
+			next:           next,
+			logger:         logger,
+			enabled:        cfg.Enabled,
+			authLevelNames: cfg.AuthLevelNames,
+		}
+	}
+}
+
+// MultiFactorAuthentication is a authenticator that checks for mfa on specific paths
+type MultiFactorAuthentication struct {
+	next           http.Handler
+	logger         log.Logger
+	enabled        bool
+	authLevelNames []string
+}
+
+// ServeHTTP adds the mfa header if the request contains a valid mfa token
+func (m MultiFactorAuthentication) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	defer m.next.ServeHTTP(w, req)
```

Comment:
Even considering code duplication, I'd move the call to were it's expected to be.
The basic pattern for middlewares is "do something before" -> "call next.ServeHTTP" -> "do something after". It isn't uncommon for middlewares to return early without calling the next one, aborting the request.

Using defer here will be confusing because the `next.ServeHTTP` is THE function everyone expect to be executed, but it's relegated to a secondary cleanup function. In addition, breaking the expected pattern will cause confusion if we need to make changes later.